File: scripts/build_hgnc_map.R
# =============================================================================
# Build HGNC gene symbol mapping (alias → official symbol)
# R equivalent of build_hgnc_map.py
#
# Downloads HGNC data and creates hgnc_map.json
# Maps: previous symbols, alias symbols, Ensembl IDs → approved symbol
#
# Usage:
#   source("build_hgnc_map.R")
#   (writes ../hgnc_map.json)
# =============================================================================

library(jsonlite)

HGNC_URL <- paste0(
  "https://www.genenames.org/cgi-bin/download/custom?",
  "col=gd_app_sym&col=gd_prev_sym&col=gd_aliases&col=gd_pub_ensembl_id",
  "&status=Approved&hgnc_dbtag=on&order_by=gd_app_sym_sort&format=text&submit=submit"
)

cat("Downloading HGNC data...\n")
tmp <- tempfile(fileext = ".tsv")
download.file(HGNC_URL, tmp, quiet = TRUE)
hgnc <- read.delim(tmp, stringsAsFactors = FALSE, check.names = FALSE)
cat(sprintf("Downloaded: %d genes\n", nrow(hgnc)))

# Build reverse mapping: alias → set of official symbols
reverse <- list()

add_mapping <- function(key, official) {
  key <- toupper(trimws(key))
  if (nchar(key) == 0 || key == official) return()
  if (is.null(reverse[[key]])) reverse[[key]] <<- character(0)
  reverse[[key]] <<- unique(c(reverse[[key]], official))
}

for (i in seq_len(nrow(hgnc))) {
  official <- toupper(trimws(hgnc$`Approved symbol`[i]))
  if (nchar(official) == 0) next

  # Previous symbols
  prev <- trimws(unlist(strsplit(hgnc$`Previous symbols`[i], ",")))
  for (p in prev) add_mapping(p, official)

  # Alias symbols
  aliases <- trimws(unlist(strsplit(hgnc$`Alias symbols`[i], ",")))
  for (a in aliases) add_mapping(a, official)

  # Ensembl ID
  ensembl <- toupper(trimws(hgnc$`Ensembl gene ID`[i]))
  if (nchar(ensembl) > 0) add_mapping(ensembl, official)
}

# Build final mapping (exclude ambiguous)
mapping <- list()
collisions <- 0
for (key in names(reverse)) {
  officials <- reverse[[key]]
  if (length(officials) == 1) {
    mapping[[key]] <- officials
  } else {
    collisions <- collisions + 1
  }
}

# Write JSON
output_path <- file.path(dirname(sys.frame(1)$ofile %||% "."), "..", "hgnc_map.json")
output_path <- normalizePath(output_path, mustWork = FALSE)
writeLines(toJSON(mapping, auto_unbox = TRUE), output_path)

cat(sprintf("\nMapping entries: %d\n", length(mapping)))
cat(sprintf("Ambiguous excluded: %d\n", collisions))
cat(sprintf("Output: %s\n", output_path))

File: scripts/build_hgnc_map.py
#!/usr/bin/env python3
"""
Download HGNC gene data and build a flat alias->official symbol mapping JSON.

Output: ../hgnc_map.json
  - Keys: previous symbols, alias symbols, ENSEMBL IDs (all uppercased)
  - Values: current approved symbol (uppercased)
  - Ambiguous aliases (mapping to multiple symbols) are excluded

Usage:
  python build_hgnc_map.py
"""

import csv
import json
import sys
import os
import io
import urllib.request

HGNC_URL = (
    "https://www.genenames.org/cgi-bin/download/custom?"
    "col=gd_app_sym&col=gd_prev_sym&col=gd_aliases&col=gd_pub_ensembl_id"
    "&status=Approved&hgnc_dbtag=on&order_by=gd_app_sym_sort&format=text&submit=submit"
)

OUTPUT_PATH = os.path.join(os.path.dirname(__file__), "..", "hgnc_map.json")


def main():
    print("Downloading HGNC data...", file=sys.stderr)
    req = urllib.request.Request(HGNC_URL, headers={"User-Agent": "RFinD-Tool/1.0"})
    with urllib.request.urlopen(req, timeout=60) as resp:
        raw = resp.read().decode("utf-8")
    print(f"Downloaded {len(raw):,} bytes", file=sys.stderr)

    reader = csv.DictReader(io.StringIO(raw), delimiter="\t")

    # Collect all alias -> set of official symbols
    from collections import defaultdict
    reverse = defaultdict(set)
    gene_count = 0

    for row in reader:
        official = row.get("Approved symbol", "").strip().upper()
        if not official:
            continue
        gene_count += 1

        # Previous symbols (comma-separated)
        for prev in row.get("Previous symbols", "").split(","):
            prev = prev.strip().upper()
            if prev and prev != official:
                reverse[prev].add(official)

        # Alias symbols (comma-separated)
        for alias in row.get("Alias symbols", "").split(","):
            alias = alias.strip().upper()
            if alias and alias != official:
                reverse[alias].add(official)

        # Ensembl gene ID
        ensembl = row.get("Ensembl gene ID", "").strip().upper()
        if ensembl and ensembl != official:
            reverse[ensembl].add(official)

    # Build mapping, excluding ambiguous aliases
    mapping = {}
    collisions = []
    for key, officials in sorted(reverse.items()):
        if len(officials) == 1:
            mapping[key] = next(iter(officials))
        else:
            collisions.append((key, sorted(officials)))

    # Write JSON
    output_path = os.path.abspath(OUTPUT_PATH)
    with open(output_path, "w") as f:
        json.dump(mapping, f, separators=(",", ":"), ensure_ascii=True)

    file_size = os.path.getsize(output_path)
    print(f"\nApproved genes: {gene_count:,}", file=sys.stderr)
    print(f"Mapping entries: {len(mapping):,}", file=sys.stderr)
    print(f"Ambiguous aliases excluded: {len(collisions):,}", file=sys.stderr)
    print(f"Output: {output_path} ({file_size:,} bytes / {file_size/1024/1024:.1f} MB)", file=sys.stderr)

    if collisions:
        print(f"\nFirst 20 ambiguous aliases:", file=sys.stderr)
        for k, v in collisions[:20]:
            print(f"  {k} -> {v}", file=sys.stderr)


if __name__ == "__main__":
    main()
